File: 008_String_to_Integer.cpp
class Solution {
public:
  int myAtoi(string str) {
    long long res(0);
    int sign(1), index(0);
    //handle white space
    while(str[index] == ' ') ++index;
    //handle sign
    if(str[index] == '-' || str[index] == '+')
      sign = str[index++] == '-'? -1:1;

    while(std::isdigit(str[index])){
      if(res > INT_MAX/10)
        return sign ==1? INT_MAX:INT_MIN;
      res = res * 10 + (str[index] - '0');
      ++ index;
    }

    res = sign * res;

    if(res>INT_MAX)
      return INT_MAX;
    if(res<INT_MIN)
      return INT_MIN;
    return res;
  }
};

File: java/008_String_to_Integer.java
public class Solution {
    public int myAtoi(String str) {
        int index = 0;
        int str_length = str.length();

        //  Remove Spaces
        while (index < str_length && str.charAt(index) == ' ') ++ index;

        if (index == str_length) return 0;

        // Handle signs
        int sign = 1;
        if (str.charAt(index) == '-' || str.charAt(index) == '+') {
            sign = str.charAt( index ) == '-' ? -1 : 1;
            ++ index;
        }

        // Convert number and avoid overflow
        long val = 0;
        while (index < str_length && Character.isDigit(str.charAt(index)) ) {
            val = val * 10 + (str.charAt(index) - '0');
            if (val > Integer.MAX_VALUE )
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            ++ index;
        }

        val *= sign;

        return (int)val;
    }
}
